Adds Schema.org schemas to pages

JakubMastalerz opened the pull request:
[Link to Ticket](https://torchbox.atlassian.net/browse/TWE-670)

### Description of Changes Made

- Added Organization schema markup to the homepage using JSON-LD
- Updated logo URL to use higher resolution `android-chrome-512x512.png` following Google's recommendations
- Implemented comprehensive test coverage for Organization JSON-LD functionality
- Added social media links (Bluesky, LinkedIn, Instagram) to the Organization schema
- Refactored JSON-LD tests to use direct template rendering for better reliability
- Cleaned up test imports and improved code organization

### How to Test

1. Visit the homepage at `/`
2. View page source and verify the Organization JSON-LD is present in the `<head>` section
3. Check that the logo URL points to `https://torchbox.com/android-chrome-512x512.png`
4. Verify social media links are included in the `sameAs` array
5. Run the test suite: `fab run-test` to ensure all tests pass
6. Validate the schema using Google's Rich Results Test tool

### Screenshots

<details>
  <summary>Expand to see more</summary>

The Organization JSON-LD appears in the page source as:

```html
<script type="application/ld+json">
{
    "@context": "https://schema.org",
    "@type": "Organization",
    "name": "Torchbox",
    "url": "https://torchbox.com/",
    "logo": "https://torchbox.com/android-chrome-512x512.png",
    "sameAs": [
        "https://bsky.app/profile/torchbox.com",
        "https://www.linkedin.com/company/torchbox",
        "https://www.instagram.com/torchboxltd/"
    ]
}
</script>
```

</details>

### MR Checklist

- [x] Add a description of your pull request and instructions for the reviewer to verify your work.
- [x] If your pull request is for a specific ticket, link to it in the description.
- [x] Stay on point and keep it small so the merge request can be easily reviewed.
- [x] Tests and linting passes.

#### Unit tests

- [x] Added
- [ ] Not required

#### Documentation

- [ ] Updated build docs
- [ ] Updated editor guidelines (See https://intranet.torchbox.com/torchbox-com-project-docs for the private link, for Torchbox employees only)
- [ ] Updated field spec (See https://intranet.torchbox.com/torchbox-com-project-docs for the private link, for Torchbox employees only)
- [x] Not required

#### Browser testing

- [ ] I have tested in the following browsers and environments (edit the list as required)
  - Latest version of Chrome on mac
  - Latest version of Firefox on mac
  - Latest version of Safari on mac
  - Safari on last two versions of iOS
  - Chrome on last two versions of Android
- [x] Not required

#### Data protection

- [x] Not relevant
- [ ] This adds new sources of PII and documents it and modifies Birdbath processors accordingly

#### Light and dark mode

- [ ] I have tested the changes in both light and dark mode
- [x] The change is not relevant to dark and light mode

#### Accessibility

- [ ] Automated WCAG 2.1 tests pass
- [ ] HTML validation passes
- [ ] Manual WCAG 2.1 tests completed
- [ ] I have tested in a screen reader
- [ ] I have tested in high-contrast mode
- [ ] Any animations removed for prefers-reduced-motion
- [x] Not required

#### Sustainability

- [ ] Images are optimised and lazy-loading used where appropriate
- [ ] SVGs have been optimised
- [ ] Performance and transfer of data considered
- [ ] If JavaScript is needed alternatives have been considered
- [x] Not required

#### Pattern library

- [ ] The pattern library component for this template displays correctly, and does not break parent templates
- [ ] The styleguide is updated if relevant
- [x] Changes are not relevant the pattern library

JakubMastalerz committed:
add organisation schema to homepage

JakubMastalerz committed:
add organisation schema to homepage

JakubMastalerz committed:
add breadcrumb schema to base template

JakubMastalerz committed:
add breadcrumb schema to base template

JakubMastalerz committed:
add blogposting schema

JakubMastalerz committed:
add blogposting schema

kevinhowbrook committed:
Fix blog JSON-LD tests to use direct template rendering

Tests were failing due to URL generation issues in test environment.
Direct template rendering bypasses Wagtail URL resolution problems
and provides more reliable test coverage for JSON-LD functionality.

kevinhowbrook committed:
Fix blog JSON-LD tests to use direct template rendering

Tests were failing due to URL generation issues in test environment.
Direct template rendering bypasses Wagtail URL resolution problems
and provides more reliable test coverage for JSON-LD functionality.

kevinhowbrook committed:
Add comprehensive Organization JSON-LD test coverage

Ensures Organization schema implementation is properly validated
with tests for structure, social links, logo URL, and template inclusion.
Prevents regression of homepage JSON-LD functionality.

kevinhowbrook committed:
Add comprehensive Organization JSON-LD test coverage

Ensures Organization schema implementation is properly validated
with tests for structure, social links, logo URL, and template inclusion.
Prevents regression of homepage JSON-LD functionality.

kevinhowbrook committed:
Use higher resolution logo for Organization schema

Switched from apple-touch-icon.png (180x180) to android-chrome-512x512.png (512x512)
for better rich results display and Google's recommended higher resolution.

Follows Google PWA standards and provides better quality for search engine rich snippets.

Standards followed:
- Google PWA Manifest requirements: https://developers.google.com/web/fundamentals/web-app-manifest
- Schema.org Organization logo recommendations: https://schema.org/Organization
- Google Rich Results guidelines: https://developers.google.com/search/docs/appearance/structured-data/logo

The 512x512px resolution exceeds Google's minimum requirement of 112x112px
and follows PWA best practices for high-quality icon display.

kevinhowbrook committed:
Use higher resolution logo for Organization schema

Switched from apple-touch-icon.png (180x180) to android-chrome-512x512.png (512x512)
for better rich results display and Google's recommended higher resolution.

Follows Google PWA standards and provides better quality for search engine rich snippets.

Standards followed:
- Google PWA Manifest requirements: https://developers.google.com/web/fundamentals/web-app-manifest
- Schema.org Organization logo recommendations: https://schema.org/Organization
- Google Rich Results guidelines: https://developers.google.com/search/docs/appearance/structured-data/logo

The 512x512px resolution exceeds Google's minimum requirement of 112x112px
and follows PWA best practices for high-quality icon display.

kevinhowbrook committed:
Clean up JSON-LD test imports

- Move all imports to top of file
- Remove redundant imports from test methods
- Improve code organization and readability

kevinhowbrook committed:
Clean up JSON-LD test imports

- Move all imports to top of file
- Remove redundant imports from test methods
- Improve code organization and readability

kevinhowbrook committed:
Simplify JSON-LD tests with helper methods

- Add _extract_jsonld_by_type() helper to eliminate repetitive JSON-LD extraction
- Add _get_organization_jsonld() helper for cleaner test methods
- Reduce test_organization_jsonld_social_links from 30+ lines to 8 lines
- All tests still pass with same coverage
- Much more maintainable and readable

kevinhowbrook committed:
Simplify JSON-LD tests with helper methods

- Add _extract_jsonld_by_type() helper to eliminate repetitive JSON-LD extraction
- Add _get_organization_jsonld() helper for cleaner test methods
- Reduce test_organization_jsonld_social_links from 30+ lines to 8 lines
- All tests still pass with same coverage
- Much more maintainable and readable

kevinhowbrook commented:
I've upted the PR to include tests and a more 'standard' icon.

<img width="590" height="447" alt="schematest" src="https://github.com/user-attachments/assets/a056cc17-6218-4c21-97a8-6bc441a48295" />

bmispelon commented on tbx/core/tests/test_jsonld.py:
```diff
@@ -0,0 +1,228 @@
+                pass
+            start_idx = end_idx + len(end_marker)
+
+        return json_scripts
```
**suggestion**: Use `bs4.BeautifulSoup` instead of string manipulation:
```
soup = BeautifulSoup(content, "html.parser")
scripts = [
    json.loads(tag.string)
    for tag in soup.find_all("script", type="application/ld+json")
]
return [script for script in scripts if script["@type"] == jsonld_type]
```

bmispelon commented on tbx/core/tests/test_jsonld.py:
```diff
@@ -0,0 +1,228 @@
+
+        # Test that logo URL is correct
+        self.assertEqual(logo_url, "https://torchbox.com/android-chrome-512x512.png")
+        self.assertTrue(logo_url.startswith("https://"), "Logo URL should be HTTPS")
```
This assertion is not needed since the previous line already guarantees the URL starts with `https://`.

kevinhowbrook replied:
834227ad83eaaf938ddd263afeb7a14c5b4815f4

bmispelon commented on tbx/project_styleguide/templates/patterns/pages/blog/blog-posting-jsonld.html:
```diff
@@ -0,0 +1,29 @@
+{% load wagtailcore_tags wagtailimages_tags %}
```
This template shows the limits of using templates to generate JSON. It seems very fragile to me:
* Everything has to use `|escapejs`
* tags like `{% if ... %}` need to go on the line above where they'd normally go to avoid extra blank lines
* Similarly you have to be careful about trailing commas.

It would be a lot more robust if we were to turn this into a template tag that would build the right `dict` in Python then use `json.dumps` to generate valid JSON. What do you think?

bmispelon reviewed (commented):
I've left some suggestions for improvements, mostly in the tests.

I also notice that you're using a lot of inline imports rather than putting them all at the top of the file as is normally the case, was there a reason for that?

kevinhowbrook committed:
Correct import style

kevinhowbrook committed:
Correct import style

kevinhowbrook committed:
use BeautifulSoup for JSON-LD parsing

kevinhowbrook committed:
use BeautifulSoup for JSON-LD parsing

kevinhowbrook committed:
Use built-in test client instead of render_to_string in JSON-LD tests

kevinhowbrook committed:
Use built-in test client instead of render_to_string in JSON-LD tests

kevinhowbrook committed:
Use assertCountEqual to test social links list directly

kevinhowbrook committed:
Use assertCountEqual to test social links list directly

kevinhowbrook commented on tbx/core/tests/test_jsonld.py:
```diff
@@ -0,0 +1,228 @@
+            "https://www.instagram.com/torchboxltd/",
+        ]
+        for link in expected_links:
+            self.assertIn(link, same_as)
```
Thanks, changed in dc2c3a56f956d368a9eae5f63b211e1c688f049d

kevinhowbrook committed:
Remove redundant test

kevinhowbrook committed:
Remove redundant test

kevinhowbrook commented on tbx/core/tests/test_jsonld.py:
```diff
@@ -0,0 +1,228 @@
+            self.assertTrue(link.startswith("http"), f"Invalid social link: {link}")
+
+        # Test that we have the expected number of social links
+        self.assertGreaterEqual(len(same_as), 3)
```
yep, removed with 2759764c5193c6bded031e4be40e86b5512a7521

kevinhowbrook committed:
Remove redundant assertions

kevinhowbrook committed:
Remove redundant assertions

kevinhowbrook committed:
Test breadcrumb JSON-LD against parsed dict instead of raw HTML

kevinhowbrook committed:
Test breadcrumb JSON-LD against parsed dict instead of raw HTML

kevinhowbrook committed:
Test against actual blog page

kevinhowbrook committed:
Test against actual blog page

kevinhowbrook commented on tbx/core/tests/test_jsonld.py:
```diff
@@ -0,0 +1,228 @@
+            template = get_template("patterns/pages/blog/blog-posting-jsonld.html")
+            self.assertIsNotNone(template)
+        except Exception as e:
+            self.fail(f"Blog posting JSON-LD template not found: {e}")
```
f772e2339aaffb0852f7ab7597d28deab79e4570

kevinhowbrook committed:
Remove redundant template existence tests

kevinhowbrook committed:
Remove redundant template existence tests

kevinhowbrook commented on tbx/core/tests/test_jsonld.py:
```diff
@@ -0,0 +1,228 @@
+            )
+            self.assertIsNotNone(template)
+        except Exception as e:
+            self.fail(f"Breadcrumb JSON-LD template not found: {e}")
```
d0550c771a2ce690a00687f271254524ccd3799a

kevinhowbrook committed:
Remove redundant test_multiple_jsonld_scripts test

kevinhowbrook committed:
Remove redundant test_multiple_jsonld_scripts test

kevinhowbrook commented on tbx/core/tests/test_jsonld.py:
```diff
@@ -0,0 +1,228 @@
+    def test_multiple_jsonld_scripts(self):
+        """Test that multiple JSON-LD scripts can be present on a page."""
+        org_data = self._get_organization_jsonld()
+        self.assertEqual(org_data["@type"], "Organization")
```
yea you're right - it's identical to test_base_template_includes_jsonld_block
removed in f11a11178ac77b28e2fe73d95213bbb21dde2f24

kevinhowbrook committed:
Build BlogPosting JSON-LD as Python dict instead of in template

kevinhowbrook committed:
Build BlogPosting JSON-LD as Python dict instead of in template

zerolab commented:
@helenb I would say so, yes

ianmeigh commented on tbx/blog/models.py:
```diff
@@ -228,6 +228,53 @@ def first_author(self):
+                "name": "Torchbox",
+                "logo": {
+                    "@type": "ImageObject",
+                    "url": f"{root_url}/apple-touch-icon.png",
```
Suggestion: Think the might have been missed when moving to the higher resolution icon in f5b769cfe7b9be7397e50cbcb0ca56f613114236

```suggestion
                    "url": "https://torchbox.com/android-chrome-512x512.png",
```

ianmeigh committed:
chore: use higher resolution logo for blog publisher org

Related to changes made in commit f5b769cfe7b9be7397e50cbcb0ca56f613114236

ianmeigh committed:
chore: fix linting by removing unused code

ianmeigh committed:
fix: include homepage as root for breadcrumb schema on all pages

Switched to using page.get_ancestors in the JSON schema template page as page.breadcrumbs returns empty for division pages, so using it in the JSON-LD template omitted the homepage and displayed the division page as the breadcrumb root instead.

helenb commented on tbx/project_styleguide/templates/patterns/navigation/components/breadcrumbs-jsonld.html:
```diff
@@ -1,15 +1,25 @@
             "@context": "https://schema.org/",
             "@type": "BreadcrumbList",
-            "itemListElement": [{% for crumb in page.breadcrumbs %}
+            "itemListElement": [{% for ancestor_page in page.get_ancestors|slice:'1:' %}
```
@ianmeigh this then means that the JSON-LD representation of the breadcrumb is different what is visible on the site. I'm not sure why we don't include the `home` link visibly on division pages, and maybe we should, but it seems wrong for the schema and visible representations to be at odds with each other.

Testing on http://localhost:8000/charity/blog/charity-email-marketing-soft-opt-in-changes/

From view source:

```

<script type="application/ld+json">
--
{
"@context": "https://schema.org/",
"@type": "BreadcrumbList",
"itemListElement": [
{
"@type": "ListItem",
"position": 1,
"name": "Home",
"item": "https://torchbox-com-staging.torchbox.dev/"
},
{
"@type": "ListItem",
"position": 2,
"name": "Charity",
"item": "https://torchbox-com-staging.torchbox.dev/charity/"
},
{
"@type": "ListItem",
"position": 3,
"name": "Blog",
"item": "https://torchbox-com-staging.torchbox.dev/charity/blog/"
}
]
}
</script>

```

Visible breadcrumb:

[Charity ](http://localhost:8000/charity/) / [Blog](http://localhost:8000/charity/blog/)

ianmeigh replied:
Hi @helenb, The JIRA ticket has been updated with more info regarding this query -  https://torchbox.atlassian.net/browse/TWE-670?focusedCommentId=198115

helenb replied:
In that case I'm happy to approve the changes in this commit. I won't add an approval tick to the whole pr unless you need me to.
